format/spec.md

```diff
@@ -1117,7 +1121,17 @@ Partition specs are serialized as a JSON object with the following fields:
 |**`spec-id`**|`JSON int`|`0`|
 |**`fields`**|`JSON list: [`<br />&nbsp;&nbsp;`<partition field JSON>,`<br />&nbsp;&nbsp;`...`<br />`]`|`[ {`<br />&nbsp;&nbsp;`"source-id": 4,`<br />&nbsp;&nbsp;`"field-id": 1000,`<br />&nbsp;&nbsp;`"name": "ts_day",`<br />&nbsp;&nbsp;`"transform": "day"`<br />`}, {`<br />&nbsp;&nbsp;`"source-id": 1,`<br />&nbsp;&nbsp;`"field-id": 1001,`<br />&nbsp;&nbsp;`"name": "id_bucket",`<br />&nbsp;&nbsp;`"transform": "bucket[16]"`<br />`} ]`|
 
-Each partition field in the fields list is stored as an object. See the table for more detail:
+Each partition field in the `fields` is stored as a JSON object with the following properties.
+
+| V1       | V2       | V3       | Field            | JSON representation | Example      |
+|----------|----------|----------|------------------|---------------------|--------------|
+| required | required | required | **`source-id`**  | `JSON int`          | 1            |
```

Comment:
I wonder whether we actually want to make `source-id` required in V3. The spec says:

```
The format version number is incremented when new features are added that will break forward-compatibility---that is, when older readers would not read newer table features correctly.
```

I understand that if we have add a compatibility flag to V2 to write this, we will have to populate `source-id` with -1. Will it be enough to simply not write `source-id` in V3 and make `source-ids` required? We can have some special instructions for how to read older metadata in V3.

Reply:
Yes, so for backward compatibility, here is the flow for newer readers:

1. if table is V3 (or V1/V2 table with a special flag), read from source-ids.
2.If table is V1/V2 without special flag, read from soruce-id.

Then write:
1. if table is V3 (or V1/V2 table with a special flag), write to source-ids.
2. If table is V1/V2 without special flag, write source-id.

For forward compatibility, an old reader will not be able to read the new tables with `source-ids` field.  They will have an error de-serializing the JSON object for partition/sort fields.  

The -1 idea is just to make it so they get further, but they will still error when they realize it is not a valid field id.